### PR TITLE
Browse models: expandable/collapsible collections

### PR DESCRIPTION
**Pull request by rafpaf**
Needs tests

This PR allows the user to expand and collapse collections in Browse data. The user can show and hide the entire collection using the chevron icon to the left of the header. Independently, the user can show all of the models within the collection, or just a portion, using the 'Show all'/'Show less' button. These toggles are independent.

The user's preferences about which collection is shown are persisted across sessions using a database setting. (Note: there is a concern about race conditions here, since the API calls that update the setting could hypothetically be handled by the backend in a different order than they were triggered in. I have left this problem unresolved for now.)

expand-collapes-colls

- [x] has tests

### CODE REVIEW

File: frontend/src/metabase/browse/components/ModelGroup.tsx [hunk 1]
```diff
@@ -1,54 +1,199 @@
...
+  const dispatch = useDispatch();
+
+  const newCollectionPrefs = {
+    ...initialCollectionPrefs,
+    [collection.id]: { areAllModelsShown, areSomeModelsShown },
+  };
+  const newCollectionPrefsStringified = JSON.stringify(newCollectionPrefs);
+
+  useEffect(() => {
+    // FIXME: Can this lead to race conditions? Perhaps a system with Promise.all would make sense?
```
Comment (rafpaf) [inline frontend/src/metabase/browse/components/ModelGroup.tsx:115]: ^ Important to discuss this

File: frontend/src/metabase/browse/components/BrowseDatabases.tsx [hunk 2]
```diff
@@ -4,7 +4,6 @@ import { color } from "metabase/lib/colors";
 import * as Urls from "metabase/lib/urls";
```
Comment (rafpaf) [inline frontend/src/metabase/browse/components/BrowseDatabases.tsx:4]: This small styling change is not related to the theme of this PR. Happy to extract it out.

File: frontend/src/metabase/browse/components/ModelGroup.tsx [hunk 3]
```diff
@@ -1,54 +1,199 @@
...
+  try {
+    const parsed = JSON.parse(json || "");
+    return isValidCollectionPrefs(parsed) ? parsed : null;
+  } catch (e) {
+    if (e instanceof SyntaxError) {
+      return null;
+    }
+    throw e;
+  }
+};
```
Comment (iethree) [inline frontend/src/metabase/browse/components/ModelGroup.tsx:71]: 1) I don't think settings like this should be server-side, they're simple UI interactions, not data we need to persist. we don't want to cause a bunch of network requests and db write/reads every time the user simply manipulates the UI.
2) We should lift the expand/collapse show/hide state up since it's all one object and only pass down necessary properties - we have potentially thousands of collections here and we don't need them all doing the exact same parsing when we can do it once at the parent-component level.
3) these helper functions are great, but also perfect candidates to be extracted to a `utils` file to keep the component easier to understand at a glance.

**Review by kdoh — APPROVED** [review]

File: frontend/src/metabase/browse/components/BrowseModels.tsx [hunk 4]
```diff
@@ -34,20 +39,65 @@ export const BrowseModels = ({
...
+                }
+                showAll={!!collectionViewPreferences?.[collectionId]?.showAll}
+                toggleExpanded={() =>
+                  handleToggleCollectionExpand(collectionId)
+                }
+                toggleShowAll={() =>
+                  handleToggleCollectionShowAll(collectionId)
+                }
+                models={groupOfModels}
+                key={`modelgroup-${groupOfModels[0].collection.id}`}
```
Comment (sloansparger) [inline frontend/src/metabase/browse/components/BrowseModels.tsx:96]: world's tiniest nit (only say this because I want us to ban index access at some point since it's not typesafe)
```suggestion
                key={`modelgroup-${collectionId}`}
```

File: frontend/src/metabase/browse/components/ModelGroup.tsx [hunk 5]
```diff
@@ -1,34 +1,59 @@
...
+
+    return {
+      aboveFoldModels: sortedModels.slice(0, MAX_COLLAPSED_MODELS),
+      belowFoldModels: sortedModels.slice(MAX_COLLAPSED_MODELS),
+    };
+  }, [models, localeCode]);
+
+  const visibleModels = showAll
+    ? [...aboveFoldModels, ...belowFoldModels]
+    : aboveFoldModels;
```
Comment (sloansparger) [inline frontend/src/metabase/browse/components/ModelGroup.tsx:56]:
```suggestion
  const sortedModels = useMemo(() => {
     return [...models].sort((a, b) => sortModels(a, b, localeCode))
  }, [models, localeCode]);
  
  const visibleModels = useMemo(() => {
    return showAll
      ? sortedModel
      : sortedModels.slice(0, MAX_COLLAPSED_MODELS)
  }, [sortedModels, showAll]);
```

I think this would be preferable for two reason:
1. sort mutates the underlying array, so it's best to copy first
2. slice shallow copies, but then we reconstruct sortedModels by doing `[...aboveFoldModels, ...belowFoldModels]`. this way we can avoid the extra copies.

File: frontend/src/metabase/browse/utils.ts [hunk 6]
```diff
@@ -136,3 +145,52 @@ export const filterModels = (
...
+
+type ModelVisibilityPrefs = {
+  expanded: boolean;
+  showAll: boolean;
+};
+
+const isRecordWithCollectionIdKeys = (
+  prefs: unknown,
+): prefs is Record<CollectionId, any> =>
+  typeof prefs === "object" &&
```
Comment (sloansparger) [inline frontend/src/metabase/browse/utils.ts:159]: We should have some better type utils. This would incorrectly work with an array.

**Review by sloansparger — APPROVED** [review]
approved with one comment
